### PR TITLE
Proxy D2L files using python

### PR DESCRIPTION
**Pull request by marcospri**
Needs https://github.com/hypothesis/h-vialib/pull/42 merged and published before CI is green.

For 
- https://github.com/hypothesis/lms/issues/4768

## Testing

Follow the instructions over: https://github.com/hypothesis/lms/pull/4792

### CODE REVIEW

File: via/requests_tools/headers.py [hunk 1]
```diff
@@ -75,11 +77,22 @@ def clean_headers(headers):
...
 
-def add_request_headers(headers):
+def add_request_headers(headers, request=None):
     """Add headers for 3rd party providers which we access data from."""
 
     # Pass our abuse policy in request headers for third-party site admins.
     headers["X-Abuse-Policy"] = "https://web.hypothes.is/abuse-policy/"
     headers["X-Complaints-To"] = "https://web.hypothes.is/report-abuse/"
 
+    if request and "via.secret.headers" in request.params:
```
Comment (marcospri) [inline via/requests_tools/headers.py:87]: Adding this here so every request benefits from it (at least the python ones).

File: via/services/pdf_url.py [hunk 2]
```diff
@@ -73,20 +75,35 @@ def get_pdf_url(self, url):
...
     @classmethod
     def _is_onedrive_url(cls, url):
         return any(regexp.match(url) for regexp in cls._MS_ONEDRIVE_URL)
 
-    def _proxy_onedrive_pdf(self, url):
+    @classmethod
+    def _is_d2l_url(cls, url):
+        return any(regexp.match(url) for regexp in cls._D2L_URL)
+
+    def _proxy_python_pdf(self, url, route="proxy_python_pdf"):
```
Comment (marcospri) [inline via/services/pdf_url.py:93]: Making this method a bit generic instead of copy/paste a similar version.

File: via/views/__init__.py [hunk 3]
```diff
@@ -21,10 +21,10 @@ def add_routes(config):  # pragma: no cover
         factory=QueryURLResource,
     )
     config.add_route(
-        "proxy_onedrive_pdf",
-        "/onedrive/proxied.pdf",
-        factory=QueryURLResource,
+        "proxy_onedrive_pdf", "/onedrive/proxied.pdf", factory=QueryURLResource
     )
+    config.add_route("proxy_d2l_pdf", "/d2l/proxied.pdf", factory=QueryURLResource)
+    config.add_route("proxy_python_pdf", "proxied.pdf", factory=QueryURLResource)
```
Comment (marcospri) [inline via/views/__init__.py:27]: This last route is a bit extra and not used but I reckon it leads to a nice "just proxy these with python" feature in the future without too much noise.

File: via/views/view_pdf.py [hunk 4]
```diff
@@ -41,11 +41,17 @@ def view_pdf(context, request):
...
+def proxy_python_pdf(context, request):
+    """Proxy a pdf with python (as opposed to nginx).
+
+    Multiple routes point to this view to allow having separate access logs for each of them.
+    """
     url = context.url_from_query()
 
     content_iterable = request.find_service(HTTPService).stream(
-        url, headers=add_request_headers({})
+        url, headers=add_request_headers({}, request=request)
```
Comment (marcospri) [inline via/views/view_pdf.py:54]: Passing the request here get the original headers.

File: via/services/pdf_url.py [hunk 5]
```diff
@@ -73,20 +75,35 @@ def get_pdf_url(self, url):
             return self._google_file_url(file_details, url)
 
         if self._is_onedrive_url(url):
-            return self._proxy_onedrive_pdf(url)
+            return self._proxy_python_pdf(url, route="proxy_onedrive_pdf")
+
+        if self._is_d2l_url(url):
```
Comment (marcospri) [inline via/services/pdf_url.py:80]: I briefly tried the nginx approach but not sure if it was possible to:

- Pass the headers to nginx in a secure way
- Do that in a way that involved sane code.

**Review by seanh — APPROVED** [review]
Not much to say, this all looks sensible.

It might be worth having a check in New Relic at some point to see how Via's performance is doing with now proxying lots of files in Python instead of NGINX.

**Pull request merged**